630157 - [6.1 FEAT] Bridge MIBs / 802.1q (vlan) support [202043]

Continuation of net-snmp-5.5-bridge-mib2.patch, beware,
three upstream commits are in this file.

commit bb1941c3cf361ab4961a04a7d0a60aa98f263718
Author: jsafranek <jsafranek@06827809-a52a-0410-b366-d66718629ded>
Date:   Wed Feb 16 13:16:57 2011 +0000

    Fixed GETNEXT processing in snmp-bridge-mib.
    
    
    git-svn-id: https://net-snmp.svn.sourceforge.net/svnroot/net-snmp/trunk@19937 06827809-a52a-0410-b366-d66718629ded

diff --git a/net-snmp/local/snmp-bridge-mib b/net-snmp/local/snmp-bridge-mib
index f6d13e6..b240492 100644
--- a/net-snmp/local/snmp-bridge-mib
+++ b/net-snmp/local/snmp-bridge-mib
@@ -123,19 +123,44 @@ sub mode_get{
 	reply($request, $noid) if (exists $oid_value{$noid});
 }
 
+sub find_next{
+	my ($noid) = @_;
+	my $nextoid = $oid_next{$noid};
+	if(!defined($nextoid) || !defined($oid_value{$nextoid})) {
+		# find the lowest OID whis is higher than $noid
+		$prev = ".1.3.6.1.2.1.18";
+		$prev_oid = new NetSNMP::OID($prev);
+		$noid_oid = new NetSNMP::OID($noid);
+		#print "looking for next of $noid\n";
+		for my $candidate (keys %oid_value) {
+			#print "evaluating $candidate\n";
+			$candidate_oid = new NetSNMP::OID($candidate);
+			if ($noid_oid < $candidate_oid && $prev_oid > $candidate_oid) {
+				#print "found candidate $candidate\n";
+				$prev = $candidate;
+				$prev_oid = $candidate_oid;
+			}
+		}
+		if ($prev eq ".1.3.6.1.2.1.18") {
+			return; # no OID higher than $noid found
+		}
+		$nextoid = $prev;
+	}
+	return $nextoid;
+}
+
 sub mode_get_next{
 	my ($request)=@_;
 	my $oid=$request->getOID();
 	$SNMP::use_numeric = 1;
 	my $noid=SNMP::translateObj($oid);
-
-	my $nextoid = $oid_next{$noid};
-	my $type = $oid_type{$nextoid};
-	my $value = $oid_value{$nextoid};
-
+	my $nextoid = find_next($noid);
+	#print "found $nextoid\n";
 	if(defined($nextoid)) {
+		my $type = $oid_type{$nextoid};
+		my $value = $oid_value{$nextoid};
 		if (defined($type) and defined($value)) {
-			reply($request,	$oid_next{$noid});
+			reply($request,	$nextoid);
 		}
 	}
 }

commit 6d1a78a2fc762792c4bca99d1ceebb64eac24705
Author: jsafranek <jsafranek@06827809-a52a-0410-b366-d66718629ded>
Date:   Wed Feb 16 14:56:33 2011 +0000

    Fixed snmp-bridge-mib when working in the embedded perl.
    
    git-svn-id: https://net-snmp.svn.sourceforge.net/svnroot/net-snmp/trunk@19938 06827809-a52a-0410-b366-d66718629ded

diff --git a/net-snmp/local/snmp-bridge-mib b/net-snmp/local/snmp-bridge-mib
index b240492..7327086 100644
--- a/net-snmp/local/snmp-bridge-mib
+++ b/net-snmp/local/snmp-bridge-mib
@@ -46,26 +46,29 @@ my %oid_next;
 my (%indexes, %interfaces, %macs, %ages, %locals, %vlans, %tagged);
 my $oid;
 
-my $subagent=0;
+$subagent=0;
 my $netdir="/sys/class/net/";
 
 $cache_timout=60;
 $last_populated=0;
 
 my $regoid = ".1.3.6.1.2.1.17";
-
-if ( $#ARGV != 0) {
-	usage($0);
-	exit();
-}
-
-my $targetbridge = $ARGV[0];
-
 # are we running embedded ? If not, register as subagent
 if (!$agent) {
     $agent = new NetSNMP::agent('Name' => 'dot1qbridge',
 				'AgentX' => 1);
     $subagent = 1;
+    if ( $#ARGV != 0) {
+	usage($0);
+	exit();
+    } 
+    my $targetbridge = $ARGV[0];
+} else {
+    if (!defined($bridge)) {
+	usage($0);
+	exit();
+    }
+    $targetbridge = $bridge;
 }
 
 $agent->register("dot1qbridge", ".1.3.6.1.2.1.17", \&request_handler) or die "registration of handler failed !\n";
@@ -85,15 +88,16 @@ if ($subagent) {
 
 sub usage {
 	my $name = shift;
-
-	print $name."\n\n";
-	print "usage: \t$name <bridge>	start snmp bridge module".
-		" for bridge <bridge>\n\n";
-	print "arguments:	bridge		name of the Linux bridge".
-		" which you want to provide info via snmp for.\n\n";
-
+	if ($subagent) {
+		print $name."\n\n";
+		print "usage: \t$name <bridge>	start snmp bridge module".
+			" for bridge <bridge>\n\n";
+		print "arguments:	bridge		name of the Linux bridge".
+			" which you want to provide info via snmp for.\n\n";
+	} else {
+		print 'usage in snmpd.conf: perl $bridge="br0"; perl do <path to $name>\n';
+	}
 	print "number of arguments given $#ARGV\n\n";
-
 }
 
 sub request_handler {

commit 25f3c2f9bba349d1505f56280fcc7d32136ca40c
Author: jsafranek <jsafranek@06827809-a52a-0410-b366-d66718629ded>
Date:   Thu Feb 17 11:28:27 2011 +0000

    Fixed snmp-bridge-mib in standalone (agentx) version.
    Don't declare local variables inside block. Patch provided by Jens Osterkamp.
    
    git-svn-id: https://net-snmp.svn.sourceforge.net/svnroot/net-snmp/trunk@19939 06827809-a52a-0410-b366-d66718629ded

diff --git a/net-snmp/local/snmp-bridge-mib b/net-snmp/local/snmp-bridge-mib
index 7327086..a4c2c80 100644
--- a/net-snmp/local/snmp-bridge-mib
+++ b/net-snmp/local/snmp-bridge-mib
@@ -48,6 +48,7 @@ my $oid;
 
 $subagent=0;
 my $netdir="/sys/class/net/";
+my $targetbridge;
 
 $cache_timout=60;
 $last_populated=0;
@@ -61,8 +62,8 @@ if (!$agent) {
     if ( $#ARGV != 0) {
 	usage($0);
 	exit();
-    } 
-    my $targetbridge = $ARGV[0];
+    }
+    $targetbridge = $ARGV[0];
 } else {
     if (!defined($bridge)) {
 	usage($0);
